### PR TITLE
Added support for container rename (datalake + blob) and container restore (datalake)

### PR DESCRIPTION
**Pull request by gapra-msft**

### CODE REVIEW

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobContainerAsyncClient.java [hunk 1]
```diff
@@ -1085,4 +1086,30 @@ private boolean validateNoETag(BlobRequestConditions modifiedRequestConditions)
...
+    Mono<Response<BlobContainerAsyncClient>> renameWithResponse(ContainerRenameOptions options, Context context) {
+        StorageImplUtils.assertNotNull("options", options);
+        BlobRequestConditions requestConditions = options.getRequestConditions() == null ? new BlobRequestConditions()
+            : options.getRequestConditions();
+        context = context == null ? Context.NONE : context;
+
+        if (!validateNoETag(requestConditions) || !validateNoTime(requestConditions)
+            || requestConditions.getTagsConditions() != null) {
+            throw logger.logExceptionAsError(new UnsupportedOperationException(
+                "Lease-Id is the only HTTP access condition supported for this API"));
```
Comment (alzimmermsft) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobContainerAsyncClient.java:1107]: If Lease-Id is the only valid conditional for this operation why do we accept the more complex match conditions object? Is this to allow for extensibility in the future?

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/options/ContainerRenameOptions.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+     */
+    public BlobRequestConditions getRequestConditions() {
+        return requestConditions;
+    }
+
+    /**
+     * @param requestConditions {@link BlobRequestConditions}
+     * @return The updated options.
+     */
+    public ContainerRenameOptions setRequestConditions(BlobRequestConditions requestConditions) {
```
Comment (alzimmermsft) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/options/ContainerRenameOptions.java:57]: Given that using options other than Lease-Id will result in a runtime exception we should add that information as a Javadoc here.

File: sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/options/FileSystemRenameOptions.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+     */
+    public DataLakeRequestConditions getRequestConditions() {
+        return requestConditions;
+    }
+
+    /**
+     * @param requestConditions {@link DataLakeRequestConditions}
+     * @return The updated options.
+     */
+    public FileSystemRenameOptions setRequestConditions(DataLakeRequestConditions requestConditions) {
```
Comment (alzimmermsft) [inline sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/options/FileSystemRenameOptions.java:57]: Will setting this with full options potentially lead to exceptions being thrown as well?

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobContainerAsyncClient.java (hunk 1 above)
Comment (gapra-msft) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobContainerAsyncClient.java:1107]: Yeah I think it is extensibility. I have seen areas in the SDK where we only take in lease id as well though so there is no set in stone pattern we've stuck with

File: sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/options/FileSystemRenameOptions.java (hunk 3 above)
Comment (gapra-msft) [reply sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/options/FileSystemRenameOptions.java:57]: Yep. I can add the exception info in the javadoc here as well

File: sdk/storage/azure-storage-file-datalake/src/test/java/com/azure/storage/file/datalake/ServiceAPITest.groovy [hunk 4]
```diff
@@ -197,4 +209,290 @@ class ServiceAPITest extends APISpec {
...
+        def blobName = generatePathName()
+        cc1.getFileClient(blobName).upload(defaultInputStream.get(), 7)
+        cc1.delete()
+        def blobContainerItem = primaryDataLakeServiceClient.listFileSystems(
+            new ListFileSystemsOptions()
+                .setPrefix(cc1.getFileSystemName())
+                .setDetails(new FileSystemListDetails().setRetrieveDeleted(true)),
+            null).first()
+
+        if (!playbackMode()) {
```
Comment (rickle-msft) [inline sdk/storage/azure-storage-file-datalake/src/test/java/com/azure/storage/file/datalake/ServiceAPITest.groovy:225]: I think at least in blobs we have a method that handles this condition

File: sdk/storage/azure-storage-file-datalake/src/test/java/com/azure/storage/file/datalake/ServiceAPITest.groovy [hunk 5]
```diff
@@ -197,4 +209,290 @@ class ServiceAPITest extends APISpec {
         response.getHeaders().getValue("x-ms-version") == "2019-02-02"
     }
 
+    def "Restore file system"() {
+        given:
+        def cc1 = primaryDataLakeServiceClient.getFileSystemClient(generateFileSystemName())
+        cc1.create()
+        def blobName = generatePathName()
+        cc1.getFileClient(blobName).upload(defaultInputStream.get(), 7)
+        cc1.delete()
```
Comment (rickle-msft) [inline sdk/storage/azure-storage-file-datalake/src/test/java/com/azure/storage/file/datalake/ServiceAPITest.groovy:218]: maybe add an exists check to ensure it actually was deleted?

File: sdk/storage/azure-storage-file-datalake/src/test/java/com/azure/storage/file/datalake/ServiceAPITest.groovy [hunk 6]
```diff
@@ -197,4 +209,290 @@ class ServiceAPITest extends APISpec {
...
+        }
+
+        when:
+        def restoredContainerClient = primaryDataLakeServiceClient.undeleteFileSystemWithResponse(
+            new FileSystemUndeleteOptions(blobContainerItem.getName(), blobContainerItem.getVersion())
+                .setDestinationFileSystemName(generateFileSystemName()), null, Context.NONE)
+            .getValue()
+
+        then:
+        restoredContainerClient.listPaths().size() == 1
```
Comment (rickle-msft) [inline sdk/storage/azure-storage-file-datalake/src/test/java/com/azure/storage/file/datalake/ServiceAPITest.groovy:262]: Validate that the restored  container has the same name as the one passed in

**Review by rickle-msft — APPROVED** [review]

File: sdk/storage/azure-storage-file-datalake/src/test/java/com/azure/storage/file/datalake/ServiceAPITest.groovy (hunk 4 above)
Comment (gapra-msft) [reply sdk/storage/azure-storage-file-datalake/src/test/java/com/azure/storage/file/datalake/ServiceAPITest.groovy:225]: Replaced with sleepIfRecord

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobContainerAsyncClient.java (hunk 1 above)
Comment (gapra-msft) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobContainerAsyncClient.java:1148]: @alzimmermsft Do you have a preference on how this gets exposed?

Comment (gapra-msft) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobContainerAsyncClient.java:1148]: We can always expose a String leaseId and deprecate the setter getter if more request conditions become valid

Comment (alzimmermsft) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobContainerAsyncClient.java:1148]: I'd lean towards using the same deciding factor as when we opt for options parameter types, if this is expected to grow in the near term future I'd lean towards keeping code as-is.

File: sdk/storage/azure-storage-file-datalake/src/test/java/com/azure/storage/file/datalake/ServiceAPITest.groovy (hunk 5 above)
Comment (gapra-msft) [reply sdk/storage/azure-storage-file-datalake/src/test/java/com/azure/storage/file/datalake/ServiceAPITest.groovy:218]: Done

File: sdk/storage/azure-storage-file-datalake/src/test/java/com/azure/storage/file/datalake/ServiceAPITest.groovy (hunk 6 above)
Comment (gapra-msft) [reply sdk/storage/azure-storage-file-datalake/src/test/java/com/azure/storage/file/datalake/ServiceAPITest.groovy:262]: Done

**Review by alzimmermsft — APPROVED** [review]

**Pull request merged**